[38942] Add non-work-package notification settings

oliverguenther opened the pull request:
Focusses mainly on [OP#38942](https://community.openproject.org/wp/38942)

Also includes [OP#38881](https://community.openproject.org/wp/38881)

oliverguenther committed:
Add new notification settings for non-wp settings

oliverguenther committed:
Add global notification setting scope

oliverguenther committed:
Turn notification settings from magic symbols into constants

oliverguenther committed:
Rewrite settings page into reactive forms

oliverguenther committed:
Change notification4 to be included in recipient's setting

oliverguenther committed:
Accept that capabilities API may return an elements: undefined result

oliverguenther committed:
Fix settings spec by checking for in_app settings

oliverguenther committed:
startWith formValue in case valueChanges did not yet fire

ulferts commented on app/services/notifications/create_from_model_service.rb:
```diff
@@ -118,78 +118,83 @@ def add_receivers_by_reason(receivers, reason)
-                        :mentioned)
+    project_applicable_settings(mentioned_ids,
+                                project,
+                                :mentioned)
```
Here, other than at the rest of the places, no constant is referenced.

oliverguenther replied:
thanks that was a relict of the rebase after your PR was merged 👍

ulferts commented on app/services/notifications/create_from_model_service/comment_strategy.rb:
```diff
@@ -50,7 +50,11 @@ def self.supports_mail?
+    User.notified_globally notification_reason(comment)
+  end
+
+  def self.notification_reason(_comment)
```
I'd rather, these methods have `subscribed` included in their names, e.g. `subscribed_reason` as they are only relevant for the subscribed case. E.g. watched is handled differently.

Since `notified_on_all` is removed from all these methods, I believe that scope can be removed. I found a reference in project.rb but I believe that references is not called any more.

oliverguenther replied:
`notified_on_all` is still referenced in the project notified_users. I remove it in the cleanup PR where the `all` setting is also removed.

ulferts replied:
:+1:

ulferts commented on app/services/notifications/create_from_model_service/work_package_strategy.rb:
```diff
@@ -30,7 +30,7 @@
 module Notifications::CreateFromModelService::WorkPackageStrategy
   def self.reasons
-    %i(mentioned involved watched subscribed commented created processed prioritized scheduled)
+    %i(mentioned involved watched commented created processed prioritized scheduled)
```
Please also remove `.subscribed_users` down below.

ulferts commented on app/models/notification_setting.rb:
```diff
@@ -1,4 +1,44 @@
+  WIKI_PAGE_UPDATED = :wiki_page_updated
+  MEMBERSHIP_ADDED = :membership_added
+  MEMBERSHIP_UPDATED = :membership_updated
+  ALL = :all
```
Will that constant then be removed again in #9665. It might not be worth the trouble then, since one can no longer configure it anyway.

oliverguenther replied:
Yes it is already removed in the cleanup PR

ulferts commented on lib/api/v3/user_preferences/notification_setting_representer.rb:
```diff
@@ -35,15 +35,23 @@ class NotificationSettingRepresenter < ::API::Decorators::Single
+        property NotificationSetting::WIKI_PAGE_UPDATED
+        property NotificationSetting::MEMBERSHIP_ADDED
+        property NotificationSetting::MEMBERSHIP_UPDATED
+        property NotificationSetting::ALL
```
You could save a couple of lines by using `NotificationSetting.all_settings.each`

ulferts replied:
Those added properties are not included in the representer spec, yet.

ulferts commented on spec/services/notifications/create_from_model_service_work_package_spec.rb:
```diff
@@ -139,7 +139,7 @@
             reason_ian: nil,
             read_mail: false,
-            reason_mail: :subscribed,
+            reason_mail: :involved,
```
As a heads up, `involved` is removed in #9683

oliverguenther committed:
Fix involved -> assigned, responsible reasons

ulferts commented on app/services/notifications/create_from_model_service/work_package_strategy.rb:
```diff
@@ -30,7 +30,7 @@
 module Notifications::CreateFromModelService::WorkPackageStrategy
   def self.reasons
-    %i(mentioned assigned responsible watched subscribed commented created processed prioritized scheduled)
+    %i(mentioned assigned responsible watched commented created processed prioritized scheduled)
```
Please remove `.subscribed_users` down below as well.

ulferts commented on frontend/src/app/features/user-preferences/reminder-settings/immediate-reminders/immediate-reminder-settings.component.html:
```diff
@@ -1,14 +1,13 @@
+                data-qa-immediate-reminder="mentioned"
+        />
+        <p slot="description">{{ text.mentioned }}</p>
+    </op-checkbox-field>
```
👍 for fixing the TODO

ulferts commented on frontend/src/app/features/user-preferences/reminder-settings/page/reminder-settings-page.component.html:
```diff
@@ -14,10 +19,16 @@ <h5 [textContent]="text.daily.title"></h5>
+    <p [textContent]="text.alerts.explanation"></p>
+    <op-email-alerts-settings></op-email-alerts-settings>
+  </section>
+  <div class="generic-table--action-buttons">
```
That class is off or named inaptly.

ulferts commented on frontend/src/app/features/user-preferences/reminder-settings/reminder-time/reminder-settings-daily-time.component.html:
```diff
@@ -1,26 +1,27 @@
       (ngModelChange)="toggleActive($event, i, selectedTimes)"
-      [disabled]="(activeTimes$ | async).length === 0 || (activeTimes$ | async).length === 1 && (activeTimes$ | async)[0] === time"
+      [ngModelOptions]="{standalone: true}"
+      [disabled]="activeTimes.length === 0 || activeTimes.length === 1 && activeTimes[0] === time"
```
Since the async pipe is no longer needed, the code could be placed inside the component.

ulferts reviewed (changes requested):
A couple of things to improve I think

## Contract

The assumption present in here, that the non work package events cannot be overwritten in a project should be manifested in the contract.

## Reminder settings page

Noticeable flickering of the checkboxes (checked/unchecked briefly) when pressing the save button. 

## Administration

Remove the now superfluous event settings:

![image](https://user-images.githubusercontent.com/617519/134007307-0221d032-7b43-429b-befe-8788e692044b.png)

oliverguenther committed:
Review feedback

oliverguenther committed:
Replace property + spec with iterator over all_settings

oliverguenther committed:
Remove flickering of values due to setLoading of query

when saving the state, the store is updated with setLoading=true,
causing the query to fire and the intermediate toggled result to be shown again.

oliverguenther committed:
Remove notified events

oliverguenther committed:
Validate a separate params contract

oliverguenther commented:
@ulferts I addressed all remarks

ulferts commented on app/workers/mails/member_job.rb:
```diff
@@ -84,8 +78,4 @@ def every_group_user_member(member, &block)
-
-  def sending_disabled?(setting, message)
-    message.blank? && !Setting.notified_events.include?("membership_#{setting}")
-  end
```
The now user specific setting for whether or not to be notified still needs to be applied here unfortunately, as the members do not yet use the notifications pipeline.

ulferts commented on modules/documents/app/controllers/documents_controller.rb:
```diff
@@ -124,14 +116,4 @@ def add_attachment
-
-    users = added.first.container.recipients
-    users.each do |user|
-      UserMailer.attachments_added(user, added).deliver_later
```
That mail is not sent out any more. 

We could extend the [DocumentStrategy](https://github.com/opf/openproject/blob/dev/modules/documents/app/services/notifications/mail_service/document_strategy.rb#L32) to check for an attachment having been added to the journal and then sending out the `attachments_added` mail.

oliverguenther replied:
I don't think that will work. This method depended on calculating which attachments were newly added. Information we don't have later on as far as I can tell. 

Since there is no setting for "attachments added to documents", I believed this could be removed. I can re-add it to the document controller sending it to all users in the project that have the global notification `documents_added` though.

oliverguenther replied:
FYI That's what I did. The documents controller now implements part of the strategies (checking allowed scope, recipients) for the attachments case. I also added some specs to ensure the emission of the mail is tested.

ulferts reviewed (changes requested):
Thanks for all the cleanup work @oliverguenther.

The flickering is gone as well. 

I fear I still have two changes that need rework, though.

The params contract approach is ok for me in this scenario. The drawback of it in my opinion is twofold:
* the client will not get all the errors in one go. In a scenario where both the timezone as well as the notification settings are wrong, the first error response will only include the notification settings errors. 
* It splits up the validations so that it becomes harder to gauge them all. 

On the other hand, it allows to not instantiate objects just for the sake of them being validated.

oliverguenther committed:
Fix check for disabling member updated notifications

oliverguenther committed:
Fix disabling of member notifications when setting is false

oliverguenther committed:
Fix document attachment mailing

oliverguenther committed:
Linting

oliverguenther committed:
Remove notified_events usage in settings

oliverguenther committed:
Remove add_attachment as no longer in use

oliverguenther committed:
Fix comments service after added param

ulferts commented on db/migrate/20210802114054_add_notification_setting_options.rb:
```diff
@@ -18,21 +17,4 @@ def add_notification_settings_options
-      end
-    end
-    # rubocop:enable Rails/WhereExists
-  end
```
The setting could now be removed from the database altogether.

oliverguenther replied:
Yes indeed :+1: As I introduce a migration in the cleanup PR, I'd suggest to remove it there instead of adding another migration here?

ulferts replied:
👍

ulferts reviewed (changes requested):
Just one minor question, @oliverguenther. :+1:
